internal/mesh/internal/controllers/gatewayproxy/fetcher/data_fetcher.go

```diff
@@ -28,6 +30,8 @@ func New(client pbresource.ResourceServiceClient, cache *cache.Cache) *Fetcher {
 }
 
 func (f *Fetcher) FetchMeshGateway(ctx context.Context, id *pbresource.ID) (*types.DecodedMeshGateway, error) {
```

Comment:
With the change here and below, I think it would be a good opportunity to add a comment on the public functions about their behavior as well as a note that they will panic if the resource id does not match the correct resource type.